parser/validate.go

```diff
@@ -242,6 +250,58 @@ func validateMessage(res *result, isProto3 bool, name protoreflect.FullName, md
 	return nil
 }
 
+func isIdentifier(s string) bool {
+	for i, r := range s {
+		if i == 0 && r >= '0' && r <= '9' {
+			// can't start with number
+			return false
+		}
+		// alphanumeric and underscore ok; everything else bad
+		switch {
+		case r >= '0' && r <= '9':
+		case r >= 'a' && r <= 'z':
+		case r >= 'A' && r <= 'Z':
+		case r == '_':
+		default:
+			return false
+		}
+	}
+	return true
+}
+
+func findMessageReservedNameNode(msgNode ast.MessageDeclNode, name string) ast.Node {
+	var decls []ast.MessageElement
+	switch msgNode := msgNode.(type) {
+	case *ast.MessageNode:
+		decls = msgNode.Decls
+	case *ast.GroupNode:
+		decls = msgNode.Decls
+	default:
+		// leave decls empty
+	}
+	return findReservedNameNode(msgNode, decls, name)
+}
+
+func findReservedNameNode[T ast.Node](parent ast.Node, decls []T, name string) ast.Node {
+	for _, decl := range decls {
+		// NB: We have to store in empty interface value first before we can do type
+		// assertion because type assertions on type parameters aren't allowed. (The
+		// compiler cannot yet know whether T is an interface type or not.)
```

Comment:
Interesting. It looks like a shorter way to write this is `any(decl).(*ast.ReservedNode)`.

Reply:
Ah, indeed. I keep forgetting about the new `any` alias for `interface{}`. Done.